identity/src/main/java/com/android/identity/util/ApplicationData.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright 2023 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.android.identity.util;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import co.nstant.in.cbor.model.Number;
+
+import java.util.Set;
+import java.util.concurrent.Callable;
+
+/**
+ * An interface to an object storing application data in a key-value pair manner.
+ *
+ * <p>This interface exists to support applications which wish to store additional data it wants to
+ * associate with a credential, for example the document type (e.g. DocType for 18013-5 credentials),
+ * user-visible name, logos/background, state, and so on.
+ */
+public interface ApplicationData {
+
+    /**
+     * Provides a set of all keys present in the ApplicationData object. This set is not backed by
+     * object, so changes to the set will not be reflected in the ApplicationData object.
+     *
+     * @return a set of all keys present in the ApplicationData object.
+     */
+    Set<String> getKeySet();
+
+    /**
+     * Allows the user to provide a {@link Callable} that they wish to be called with every change
+     * to the ApplicationData.
+     *
+     * @param saveFunc The function which will be called with every change to the ApplicationData.
+     */
+    void setSaveFunc(Callable<Boolean> saveFunc);
+
+    /**
+     * Sets application specific data.
+     *
+     * <p>Use {@link #getApplicationData(String)} to read the data back.
+     *
+     * @param key the key for the data.
+     * @param value the value or {@code null} to remove.
+     */
+    void setApplicationData(@NonNull String key, @Nullable byte[] value);
```

Comment:
I think we should just call this `setData()` since user classes (e.g. `Credential`) won't implement the interface and instead use instances of it by reference. For the specific types, I think it would also be nice to spell them out e.g. `setString()` since it makes it easier to read. Ditto naming for `get()`, `getString()`, etc.

Also would be handy for setters to return a reference to `ApplicationData` so it's possible to chain stuff, e.g.

```
Credential credential;
credential.getApplicationData()
  .setData("foo1", someBlob)
  .setString("foo2", someString);
```